### PR TITLE
Bunch of updates & fixes

### LINKED ISSUE #45: Telegram Albums Order in Discord
Telegram has now released the Album update to desktops as well, something they rolled out to mobile clients a while ago

I've noticed the pictures work just fine when TediCross sends the album pictures to Discord as individual messages, but the only issue is that they are sent out of order

So if it is doable to acquire the order of photos in an album

"As a discord user I'd like Telegram Album Photos to be sent to Discord in the same order as in the Telegram Album"

### LINKED ISSUE #181: Strikethrough formatting isn't transferred by the bridge
Strikethrough formatting is lost on messages transferred by the bridge, this affects both directions: Discord -> Telegram and Telegram -> Discord

### LINKED ISSUE #205: [Request] Spoiler and Reaction Support
Telegram has a new Feature
https://telegram.org/blog/reactions-spoilers-translations

Would be cool if the bot could do that

### LINKED ISSUE #243: Missing support for the new Telegram topics?
I set up a bridge bot for a telegram group that had topics on, producing this:

2022-12-06 00:27:38 [ERR] [REDACTED] Failed message: TelegramError: 400: Bad Request: TOPIC_DELETED
    at Telegram.callApi (/home/REDACTED/git-repos/TediCross/node_modules/telegraf/lib/core/network/client.js:264:19)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async /home/REDACTED/git-repos/TediCross/dist/discord2telegram/setup.js:205:47 {
  response: {
    ok: false,
    error_code: 400,
    description: 'Bad Request: TOPIC_DELETED'
  },
  on: {
    method: 'sendMessage',
    payload: {
      chat_id: REDACTED,
      text: '<b>REDACTED</b>\ntest',
      parse_mode: 'HTML'
    }
  }
}

I turned the topics off and lo and behold the bridge bot then worked, so I'm guessing we need an update that allows the bridge to work with telegram topics?

### LINKED ISSUE #268: Support for groups with topics enabled

### LINKED ISSUE #269: [Request] Hyperlink support
If it's posible. can you make inline link support if sending from Telegram to Discord? It seems to be supported by Discord https://www.reddit.com/r/discordapp/comments/jy968x/you_can_now_create_hyperlinks_through_messages/

### LINKED ISSUE #293: [Request] If there no @nickname then send name
When using the sendUsernames: true function in telegram, those users who do not have a username are sent simply as **, which makes it difficult to understand who is in the conversation at all. 
image

### LINKED ISSUE #306: Pins detected as message, but empty
image

### LINKED ISSUE #316: File is too big to handle
I propose to introduce a parameter that would disable these notifications, since they are generally obvious and as an alternative to add the ability to compress files to be able to transfer them.

P. S. I mean at least for video files which takes less than 50 mb

Screenshot_20230628-083329

### PR DESCRIPTION
**Pull request by Exzender**
## Changes

1. Updated to work with latest packages
2. New param (in settings -> telegram: suppressThisIsPrivateBotMessage: true) - disabling warning messages in groups with no configured bridges
   Same discord param (in settings -> discord: suppressThisIsPrivateBotMessage: true) - disabling warning messages in groups with no configured bridges
3. (#316) Option to disable "too big file" message (in settings -> telegram: suppressFileTooBigMessages: true)
3.1. Fixed: if one of files in a message is too big - the rest files text still being rerouted
3.2. Fixed: names of files sent from telegram
4. (#306) Disabled pinned message notification from TG
5. Added: Discord msg limit = 2000 chars. Use embeds for bigger messages.
6. (#293) User first & last name form Telegram used when no username set
7. (#268; #243) Added Threads support (check Readme for details)
8. (#269) For text with inline links in text use embeds.
9. Fixed: crash on adding bot to new group
10. (#181) Fixed: Strikethrough formatting
11. Fixed: if one attachment is too big other attachments still processed
12. Fixed: crash on deleting the chat with not configured bot
13. Fixed: correct parsing of "spoiler" formatting
14. Fixed: removed unnecessary dev logs
15. Package.json updated - added new packages (required for Strikethrough fix)
16. New options added to default config file (example.settings.yaml)
17. Added: photos (and other file types) transferred as groups (albums) both directions
18. Fixed: lost caption if one file in album failed on transfer
19. Fixed: lost reply from Discord to non-existing TG message
20. Fixed: crash on system message reply in discord (sync code from original TediCross repo)

## Maintainer changes

- Supersedes #317
- Feature: `/threadinfo` and `/chatinfo` are now discord slash commands and telegram chat commands

### CODE REVIEW

**Comment by Lulalaby** [comment]
This is truly amazing. Many thanks in advance.

Looking forward to review it when you mark it as ready.

Really appreciating it ❤️

File: src/bridgestuff/Bridge.ts [hunk 1]
```diff
@@ -56,14 +63,16 @@ export class Bridge {
...
-		}
+		// NOTE: redundant checks
+		// if (!(settings instanceof Object)) {
+		// 	throw new Error("`settings` must be an object");
+		// }
+		//
+		// // Check the name
+		// if (typeof settings.name !== "string") {
+		// 	throw new Error("`settings.name` must be a string");
+		// }
```
Comment (Lulalaby) [inline src/bridgestuff/Bridge.ts]: Feel free to remove completely

File: src/bridgestuff/BridgeSettingsTelegram.ts [hunk 2]
```diff
@@ -57,9 +57,11 @@ export class BridgeSettingsTelegram {
...
 	static validate(settings: BridgeSettingsTelegramProperties) {
 		// Check that the settings are indeed in object form
-		if (!(settings instanceof Object)) {
-			throw new Error("`settings` must be an object");
-		}
+
+		// NOTE: redundant check
+		// if (!(settings instanceof Object)) {
+		// 	throw new Error("`settings` must be an object");
+		// }
```
Comment (Lulalaby) [inline src/bridgestuff/BridgeSettingsTelegram.ts]: See other review

File: src/discord2telegram/setup.ts [hunk 3]
```diff
@@ -180,108 +209,169 @@ export function setup(
...
 
 					// Convert it to something Telegram likes
 					const text = handleEmbed(embed, senderName, settings.telegram);
 
 					try {
 						// Send it
-						if (replyId === "0" || replyId === undefined) {
-							tgBot.telegram.sendMessage(bridge.telegram.chatId, text, {
-								parse_mode: "HTML",
-								disable_web_page_preview: true
```
Comment (Lulalaby) [inline src/discord2telegram/setup.ts:232]: `disable_web_page_preview`: Possible candidate to become setting option

File: src/settings/DiscordSettings.ts [hunk 4]
```diff
@@ -79,15 +82,16 @@ export class DiscordSettings {
...
+		// NOTE: redundant checks
+		// // Check that the settings are indeed in object form
+		// if (!(settings instanceof Object)) {
+		// 	throw new Error("`settings` must be an object");
+		// }
+		//
+		// // Check that the token is a string
+		// if (typeof settings.token !== "string") {
+		// 	throw new Error("`settings.token` must be a string");
+		// }
```
Comment (Lulalaby) [inline src/settings/DiscordSettings.ts]: See first review

File: src/settings/Settings.ts [hunk 5]
```diff
@@ -136,9 +136,10 @@ export class Settings {
 		];
 
 		// Check that the settings are indeed in object form
-		if (!(settings instanceof Object)) {
-			throw new Error("`settings` must be an object");
-		}
+		// NOTE: redundant check
+		// if (!(settings instanceof Object)) {
+		// 	throw new Error("`settings` must be an object");
+		// }
```
Comment (Lulalaby) [inline src/settings/Settings.ts]: Top

File: src/settings/TelegramSettings.ts [hunk 6]
```diff
@@ -103,14 +114,15 @@ export class TelegramSettings {
...
-		}
+		// NOTE: redundant checks
+		// if (!(settings instanceof Object)) {
+		// 	throw new Error("`settings` must be an object");
+		// }
+		//
+		// // Check that the token is a string
+		// if (typeof settings.token !== "string") {
+		// 	throw new Error("`settings.token` must be a string");
+		// }
```
Comment (Lulalaby) [inline src/settings/TelegramSettings.ts]: Same

**Review by Lulalaby — COMMENTED** [review]
Looks good so far, just a few thing that can be removed / added.

Full review pending till I have the chance to get on my pc.

**Comment by Lulalaby** [comment]
Is https://github.com/TediCross/TediCross/issues/262 possible fixed through this PR as well?

**Comment by Exzender** [comment]
> Is #262 possible fixed through this PR as well?

this wasn't fixed - I'll add it to my todo

**Comment by Lulalaby** [comment]
Many thanks

**Comment by Fabian42** [comment]
Do I understand it correctly that all threads need to be specified in the settings manually? Is there no better way with the APIs? This would mean that chats that use many threads (like help chats) are still basically impossible to bridge.

**Comment by Lulalaby** [comment]
> I:\Development\GitHub\TediCross\src\telegram2discord\middlewares.ts:342
>                ? ctx.tediCross.message?.message_thread_id !== ctx.tediCross.message?.reply_to_message.message_id
>                                                                                        ^
> TypeError: Cannot read properties of undefined (reading 'message_id')
    at addReplyObj (I:\Development\GitHub\TediCross\src\telegram2discord\middlewares.ts:342:89)
    at execute (I:\Development\GitHub\TediCross\node_modules\telegraf\lib\composer.js:485:23)
    at I:\Development\GitHub\TediCross\node_modules\telegraf\lib\composer.js:486:27
    at execute (I:\Development\GitHub\TediCross\node_modules\telegraf\lib\composer.js:485:23)
    at I:\Development\GitHub\TediCross\node_modules\telegraf\lib\composer.js:486:27
    at addFromObj (I:\Development\GitHub\TediCross\src\telegram2discord\middlewares.ts:327:2)
    at execute (I:\Development\GitHub\TediCross\node_modules\telegraf\lib\composer.js:485:23)
    at I:\Development\GitHub\TediCross\node_modules\telegraf\lib\composer.js:486:27
    at execute (I:\Development\GitHub\TediCross\node_modules\telegraf\lib\composer.js:485:23)
    at I:\Development\GitHub\TediCross\node_modules\telegraf\lib\composer.js:486:27
[nodemon] app crashed - waiting for file changes before starting...


Creating chat topics and sending a message crashes the bot completely.

**Comment by Lulalaby** [comment]
Deletion from a telegram topic/thread won't delete from the discord thread

Rest works as intended

**Comment by Exzender** [comment]
> Do I understand it correctly that all threads need to be specified in the settings manually? Is there no better way with the APIs? This would mean that chats that use many threads (like help chats) are still basically impossible to bridge.

Topics management is possible through APIs, but I'm not sure yet how to correctly handle it (especially when they are created while the bot is offline).

**Comment by Exzender** [comment]
> Deletion from a telegram topic/thread won't delete from the discord thread
> 
> Rest works as intended

ok. thanks.

**Comment by Exzender** [comment]
> Creating chat topics and sending a message crashes the bot completely.

Yes. Topics support is in early experimental state. And deletion/creation was not handled at all.... :(

**Comment by Lulalaby** [comment]
> 

I'd say we go with your current idea..

We should ignore unknown threads (like the error) tho

**Comment by Lulalaby** [comment]
I reworked the chatinfo and threadinfo commands btw and both are now registered as discord slash command and telegram bot command

**Comment by Lulalaby** [comment]
*@Exzender could you test my changes on your side as well please?

**Comment by Exzender** [comment]
> *@Exzender could you test my changes on your side as well please?

chatinfo & threadinfo now visible in slash commands and work well

File: src/telegram2discord/From.ts [hunk 7]
```diff
@@ -79,9 +79,16 @@ export function createFromObjFromChat(chat: Record<string, any>) {
...
  */
 export function makeDisplayName(useFirstNameInsteadOfUsername: boolean, from: From) {
-	return R.ifElse<any, any, any>(
-		from => useFirstNameInsteadOfUsername || R.isNil(from.username),
-		R.prop("firstName"),
-		R.prop("username")
-	)(from);
+	if (useFirstNameInsteadOfUsername || !from.username) {
+		const prefix: string = from.lastName ? ` ${from.lastName}` : "";
+		return `${from.firstName}${prefix}`;
```
Comment (Lulalaby) [inline src/telegram2discord/From.ts]:
```suggestion
		const suffix: string = from.lastName ? ` ${from.lastName}` : "";
		return `${from.firstName}${suffix}`;
```

it's actually a suffix in this context

**Review by Lulalaby — CHANGES_REQUESTED** [review]

**Comment by Lulalaby** [comment]
Do you think #73 is fixed through point 17?

**Comment by Exzender** [comment]
> Do you think #73 is fixed through point 17?

I think yes.

**Review by Lulalaby — APPROVED** [review]

**Review by Lulalaby — APPROVED** [review]
Okay, I've tested all through. LGTM, gonna merge this.

Thank you for your huge contribution. I'm gonna wait a bit tho before syncing it to stable

**Pull request merged**